### ISSUE TITLE
UnusedLocalVariable: jdkVersion pre-22 should allow unused named pattern matched destructured variables via `instanceof` for pre-Java 22 code

### ISSUE THREAD

**Issue by james-baker-aera** [checkstyle/checkstyle#20490]
This is a follow up to #20357

---

I have read check documentation: https://checkstyle.sourceforge.io/checks/coding/unusedlocalvariable.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac -version
javac 21.0.11
```
```bash
/var/tmp $ javac Example.java
```
```bash
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="UnusedLocalVariable">
      <property name="allowUnnamedVariables" value="false"/>
      <property name="jdkVersion" value="21"/>
    </module>
  </module>
</module>
```
```bash
/var/tmp $ cat Example.java 
class Example {

    sealed interface Customer {}

    record Person(String name, int age) implements Customer {}
    record Company(String name) implements Customer {}

    String name(Customer customer) {
        if (customer instanceof Person(String name, int ignoredAge)) {
            return name;
        } else if (customer instanceof Company(String name)) {
            return name;
        }
        throw new IllegalStateException();
    }
}
```
```bash
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
```
```bash
/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.7.0-all.jar -c config.xml Example.java
Starting audit...
[ERROR] /var/tmp/Example.java:9:57: Unused local variable 'ignoredAge'. [UnusedLocalVariable]
Audit done.
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

In pre-Java 22 code you can't write `customer instanceof Person(String name, int _)`, so the unused variable must be present. When the `jdkVersion` of the `UnusedLocalVariable` check is set to a pre-Java 22 version, this should not error.

---

Note that this is fine for `switch` statements:

```bash
/var/tmp $ javac AnotherExample.java
```
```bash
/var/tmp $ cat AnotherExample.java 
import java.util.stream.Stream;

class AnotherExample {

    sealed interface Customer {}

    record Person(String name, int age) implements Customer {}
    record Company(String name) implements Customer {}

    String name(Customer customer) {
        return switch (customer) {
            case Person(String name, int ignoredAge) -> name;
            case Company(String name) -> name;
        };
    }
}
```
```bash
/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.7.0-all.jar -c config.xml AnotherExample.java 
Starting audit...
Audit done.
```

**Comment by romani** [comment]
@ZaheerAhmadDev , please help to extend fix for this Check.

**Comment by ZaheerAhmadDev** [comment]
IMO: I think the alternative would be to avoid destructuring entirely:
```java
class Example {

    sealed interface Customer {}

    record Person(String name, int age) implements Customer {}

    String name(Customer customer) {
        if (customer instanceof Person person) {
            return person.name();
            }
        throw new IllegalStateException();
    }
}
```
```
$ java  -jar checkstyle-13.7.0-all.jar -c config.xml src/Example.java
Starting audit...
Audit done.
```
@romani Both approaches produce the same result. Should we suppress the violation for instanceof record destructuring when `jdkVersion < 22`, or should the user be expected to avoid record destructuring entirely?

**Comment by ZaheerAhmadDev** [comment]
For Switch-case user on `pre-java 22` has no alternative compiler forces a name pattern.But for `instanceof record` destructuring the user does have an alternative **avoid destructuring entirely**.

**Comment by romani** [comment]
@ZaheerAhmadDev , I am not syure what you mean, please share code snippents

**Comment by ZaheerAhmadDev** [comment]
We have an option for `instance of` if we dont want to use variable.
**Option A- destructuring with an Unused component (Violation reported):**
```java
class Example {

    sealed interface Customer {}

    record Person(String name, int age) implements Customer {}
    record Company(String name) implements Customer {}

    String name(Customer customer) {
        if (customer instanceof Person(String name, int ignoredAge)) {
            return name;
        } else if (customer instanceof Company(String name)) {
            return name;
        }
        throw new IllegalStateException();
    }
}

$ java  -jar checkstyle-13.7.0-all.jar -c config.xml src/Example.java
Starting audit...
[ERROR] C:\Users\PMYLS\Desktop\check\src\Example.java:9:57: Unused local variable 'ignoredAge'. [UnusedLocalVariable]
Audit done.
Checkstyle ends with 1 errors.
```

**Option B- don't destructure the field you dont need (no violation work on any jdk version):**
```java
class Example { 
    sealed interface Customer {}

    record Person(String name, int age) implements Customer {}

    String name(Customer customer) {
        if (customer instanceof Person person) {
            return person.name();
        }
        throw new IllegalStateException();
    }
}

$ java  -jar checkstyle-13.7.0-all.jar -c config.xml src/Example.java
Starting audit...
Audit done.
```
If `age` isn't needed, there's no reason to destructure it out at all  option B gives identical behavior with zero unused variables, on any jdk version. The unused variable in option A only exists because the user chose to destructure a field they didn't need, not because of a jdk limitation. That's why I don't think this case should be suppressed by `jdkVersion`.

**Comment by ZaheerAhmadDev** [comment]
@romani @james-baker-aera @mahfouz72  please share your opinion on this, am i missing a point?

**Comment by romani** [comment]
> `In pre-Java 22 code you can't write customer instanceof Person(String name, int _)`

is this compie-able syntax in jdk22/25 ?

**Comment by james-baker-aera** [comment]
> > `In pre-Java 22 code you can't write customer instanceof Person(String name, int _)`
> 
> is this compie-able syntax in jdk22/25 ?

Yes

https://docs.oracle.com/en/java/javase/22/language/unnamed-variables-and-patterns.html

**Comment by james-baker-aera** [comment]
@ZaheerAhmadDev I agree there are workarounds.

My main issues are:
1. This is a regression (it passes on Checkstyle `13.5.0`) – so it would force us to rewrite our code to upgrade. 
2. There are other use-cases for nested record pattern matching
   ```java
   sealed interface Maybe<T> {};

   record None<T>() implements Maybe<T> {};
   record Some<T>(T value) implements Maybe<T> {};
   ```
   You might want code like this in a function 
   ```java
   boolean isNested(Maybe<?> maybe) {
       if (maybe intanceof Some(Some inner) { // You need the variable `inner` here
           return true;
       }
       return false;
   }
   ```

**Comment by mahfouz72** [comment]
I agree that this is issue and need to be solved, destructured instance of record patterns needs to be skipped pre jdk-22

**Comment by ZaheerAhmadDev** [comment]
i am on it.

**Comment by romani** [comment]
fix is merged.
All gratitude to @ZaheerAhmadDev  !!! 

I will do release tomorrow.
